packages/vm/src/instance.rs

```diff
@@ -181,13 +189,127 @@ where
 
 #[cfg(test)]
 mod test {
-    use crate::calls::{call_handle, call_init, call_query};
-    use crate::testing::{mock_instance, mock_instance_with_gas_limit};
+    use super::*;
+
     use cosmwasm_std::coin;
     use cosmwasm_std::testing::mock_env;
+    use wasmer_runtime_core::error::ResolveError;
+
+    use crate::calls::{call_handle, call_init, call_query};
+    use crate::errors::Error;
+    use crate::testing::{mock_instance, mock_instance_with_gas_limit};
 
+    static KIB: usize = 1024;
+    static MIB: usize = 1024 * 1024;
     static CONTRACT: &[u8] = include_bytes!("../testdata/contract.wasm");
 
+    #[test]
+    fn func_works() {
+        let instance = mock_instance(&CONTRACT);
+
+        // can get func
+        let allocate: Func<u32, u32> = instance.func("allocate").expect("error getting func");
+
+        // can call a few times
+        let _ptr1 = allocate.call(0).expect("error calling allocate func");
+        let _ptr2 = allocate.call(1).expect("error calling allocate func");
+        let _ptr3 = allocate.call(33).expect("error calling allocate func");
+    }
+
+    #[test]
+    fn func_errors_for_non_existent_function() {
+        let instance = mock_instance(&CONTRACT);
+        let missing_function = "bar_foo345";
+        match instance.func::<(), ()>(missing_function) {
+            Err(Error::ResolveErr { source, .. }) => match source {
+                ResolveError::ExportNotFound { name } => assert_eq!(name, missing_function),
+                _ => panic!("found unexpected source error"),
+            },
+            Err(e) => panic!("unexpected error: {:?}", e),
+            Ok(_) => panic!("must not succeed"),
+        }
+    }
+
+    #[test]
+    fn allocate_deallocate_works() {
+        let mut instance = mock_instance(&CONTRACT);
+
+        let sizes: Vec<usize> = vec![
+            0,
+            4,
+            40,
+            400,
+            4 * KIB,
+            40 * KIB,
+            400 * KIB,
+            4 * MIB,
+            40 * MIB,
+            400 * MIB,
+        ];
+        for size in sizes.into_iter() {
+            let region_ptr = instance.allocate(size).expect("error allocating");
+            instance.deallocate(region_ptr).expect("error deallocating");
+        }
+    }
+
+    #[test]
+    fn write_and_read_memory_works() {
+        let mut instance = mock_instance(&CONTRACT);
+
+        let sizes: Vec<usize> = vec![
+            0,
+            4,
+            40,
+            400,
+            4 * KIB,
+            40 * KIB,
+            400 * KIB,
+            4 * MIB,
+            // disabled for performance reasons, but pass as well
+            // 40 * MIB,
+            // 400 * MIB,
+        ];
+        for size in sizes.into_iter() {
+            let region_ptr = instance.allocate(size).expect("error allocating");
+            let original = vec![0u8; size];
```

Comment:
It would be interesting to generate some non-0 data (to make it clear this isn't just an artifact of initializing memory. Even `vec![1u8; size]` would be more clearly writen.

Reply:
Done in https://github.com/CosmWasm/cosmwasm/pull/215/commits/3791c78886a78a17e959bcf12a35c2059894677c